File: 1242/main.cs
using System;
using System.Collections.Generic;

class Program
{
    static void FindWerewolfSuspects(int n, List<Tuple<int, int>> relationships, HashSet<int> victims)
    {
        var ancestors = new Dictionary<int, HashSet<int>>();
        var children = new Dictionary<int, HashSet<int>>();

        // Build the family tree
        foreach (var relationship in relationships)
        {
            int child = relationship.Item1;
            int parent = relationship.Item2;
            if (!children.ContainsKey(parent)) children[parent] = new HashSet<int>();
            children[parent].Add(child);

            if (!ancestors.ContainsKey(child)) ancestors[child] = new HashSet<int>();
            ancestors[child].Add(parent);
            if (ancestors.ContainsKey(parent))
            {
                foreach (var ancestor in ancestors[parent])
                {
                    ancestors[child].Add(ancestor);
                }
            }
        }

        // Find all ancestors of victims
        var victimAncestors = new HashSet<int>();
        foreach (var victim in victims)
        {
            if (ancestors.ContainsKey(victim))
            {
                foreach (var ancestor in ancestors[victim])
                {
                    victimAncestors.Add(ancestor);
                }
            }
        }

        // Determine suspects (those who are not ancestors of the victims)
        var suspects = new HashSet<int>();
        for (int i = 1; i <= n; i++)
        {
            if (!victimAncestors.Contains(i) && !victims.Contains(i))
            {
                suspects.Add(i);
            }
        }

        // Print suspects
        if (suspects.Count == 0)
        {
            Console.WriteLine("0");
        }
        else
        {
            var sortedSuspects = new List<int>(suspects);
            sortedSuspects.Sort();
            Console.WriteLine(string.Join(" ", sortedSuspects));
        }
    }

    static void Main(string[] args)
    {
        int n = int.Parse(Console.ReadLine());
        var relationships = new List<Tuple<int, int>>();
        string line;

        // Reading relationships
        while ((line = Console.ReadLine()) != "BLOOD")
        {
            var parts = line.Split();
            int child = int.Parse(parts[0]);
            int parent = int.Parse(parts[1]);
            relationships.Add(new Tuple<int, int>(child, parent));
        }

        // Reading victims
        var victims = new HashSet<int>();
        while ((line = Console.ReadLine()) != null && line.Trim() != "")
        {
            victims.Add(int.Parse(line.Trim()));
        }

        FindWerewolfSuspects(n, relationships, victims);
    }
}

File: 1242/main.cpp
#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <sstream>
#include <algorithm>

using namespace std;

void find_werewolf_suspects(int n, const vector<pair<int, int>> &relationships, const set<int> &victims)
{
    map<int, set<int>> ancestors;
    map<int, set<int>> children;

    // Build the family tree
    for (const auto &relationship : relationships)
    {
        int child = relationship.first;
        int parent = relationship.second;
        children[parent].insert(child);
        ancestors[child].insert(parent);
        ancestors[child].insert(ancestors[parent].begin(), ancestors[parent].end());
    }

    // Find all ancestors of victims
    set<int> victim_ancestors;
    for (int victim : victims)
    {
        victim_ancestors.insert(ancestors[victim].begin(), ancestors[victim].end());
    }

    // Determine suspects (those who are not ancestors of the victims)
    set<int> suspects;
    for (int i = 1; i <= n; ++i)
    {
        if (victim_ancestors.find(i) == victim_ancestors.end() && victims.find(i) == victims.end())
        {
            suspects.insert(i);
        }
    }

    // Print suspects
    if (suspects.empty())
    {
        cout << "0" << endl;
    }
    else
    {
        for (int suspect : suspects)
        {
            cout << suspect << " ";
        }
        cout << endl;
    }
}

int main()
{
    int n;
    cin >> n;
    cin.ignore();

    vector<pair<int, int>> relationships;
    string line;
    while (getline(cin, line) && line != "BLOOD")
    {
        stringstream ss(line);
        int child, parent;
        ss >> child >> parent;
        relationships.push_back({child, parent});
    }

    set<int> victims;
    while (getline(cin, line))
    {
        if (line.empty())
            break;
        victims.insert(stoi(line));
    }

    find_werewolf_suspects(n, relationships, victims);
    return 0;
}
